File: shared-domain/src/commonMain/kotlin/com/dualmusic/domain/model/Live.kt
package com.dualmusic.domain.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Live spontané d'un artiste (élément du feed vertical).
 *
 * Aligné sur `artist_lives` + l'`artist` hydraté par le backend. `roomId` est le nom de
 * room LiveKit à rejoindre ; `id` sert au chat/cadeaux (contexte `live:id`).
 */
@Serializable
data class Live(
    val id: String,
    val title: String? = null,
    val status: EventStatus = EventStatus.LIVE,
    @SerialName("artist_id") val artistId: String,
    @SerialName("room_id") val roomId: String? = null,
    @SerialName("viewer_count") val viewerCount: Int = 0,
    @SerialName("recording_url") val recordingUrl: String? = null,
    /** Profil d'affichage du host (nom, avatar, nom de scène). */
    val artist: DisplayProfile? = null,
    /** Dédicaces activées pour CE live (défaut true côté backend). */
    @SerialName("allows_dedications") val allowsDedications: Boolean = true,
    /** Prix minimum propre à CE live (crédits) — `null` = utilise le défaut global de la plateforme. */
    @SerialName("dedication_min_price_credits") val dedicationMinPriceCredits: Double? = null,
    /** Demandes d'invité (« lever la main ») activées pour CE live (défaut true côté backend). */
    @SerialName("allow_guests") val allowGuests: Boolean = true,
    /** L'artiste (hôte) a-t-il activé le chat pour ce live (défaut true côté backend). */
    @SerialName("chat_enabled") @Serializable(with = com.dualmusic.domain.serialization.FlexibleBoolSerializer::class) val chatEnabled: Boolean = true,
) {
    /**
     * Room LiveKit effective. DOIT correspondre exactement à la convention du web
     * (`live-<id>`, cf. LiveStream.tsx) pour que web et mobile publient/regardent la
     * MÊME room. Le backend laisse `room_id` à null → on dérive toujours `live-<id>`.
     */
    val liveKitRoom: String get() = "live-$id"
}

File: ios/Packages/DualMusicKit/Sources/DomainModels/Live.swift
import Foundation

/// Live spontané d'un artiste (élément du feed vertical).
///
/// Aligné sur la table `artist_lives` + l'`artist` hydraté par le backend. ``liveKitRoom``
/// est le nom de room LiveKit à rejoindre ; ``id`` sert au chat et aux cadeaux (contexte
/// `live:id`). Miroir de `shared-domain/model/Live.kt`.
public struct Live: Codable, Sendable, Identifiable, Equatable {
    public let id: String
    public let title: String?
    public let status: EventStatus
    public let artistId: String
    public let roomId: String?
    public let viewerCount: Int
    public let recordingURL: String?
    /// Dédicaces activées pour ce live (réglage hôte, réactif en direct via l'événement
    /// temps réel `settings`).
    public let allowsDedications: Bool
    /// Prix minimum d'une dédicace propre à CE live (`nil` = défaut global,
    /// `GET /settings/public/economic_config`).
    public let dedicationMinPriceCredits: Double?
    /// Demandes d'invité (« lever la main ») activées pour ce live.
    public let allowGuests: Bool
    /// Chat activé pour ce live (jamais délégué à un modérateur, réservé à l'hôte).
    public let chatEnabled: Bool
    /// Profil d'affichage du host (nom, avatar, nom de scène).
    public let artist: DisplayProfile?

    enum CodingKeys: String, CodingKey {
        case id, title, status
        case artistId = "artist_id"
        case roomId = "room_id"
        case viewerCount = "viewer_count"
        case recordingURL = "recording_url"
        case allowsDedications = "allows_dedications"
        case dedicationMinPriceCredits = "dedication_min_price_credits"
        case allowGuests = "allow_guests"
        case chatEnabled = "chat_enabled"
        case artist
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = c.val(String.self, .id, "")
        title = c.opt(String.self, .title)
        status = c.val(EventStatus.self, .status, .live)
        artistId = c.val(String.self, .artistId, "")
        roomId = c.opt(String.self, .roomId)
        viewerCount = c.int(.viewerCount)
        recordingURL = c.opt(String.self, .recordingURL)
        allowsDedications = c.bool(.allowsDedications, true)
        dedicationMinPriceCredits = c.amountIfPresent(.dedicationMinPriceCredits)
        allowGuests = c.bool(.allowGuests, true)
        chatEnabled = c.bool(.chatEnabled, true)
        artist = c.opt(DisplayProfile.self, .artist)
    }

    /// Room LiveKit effective (repli sur `live-id` si `room_id` est absent).
    ///
    /// ⚠️ Le tiret compte : le backend/Android dérivent `"live-$id"` (`Live.kt`) — le repli
    /// utilisait `"live:\(id)"` (deux-points) avant ce correctif, ce qui aurait empêché de
    /// rejoindre la bonne room LiveKit pour tout live sans `room_id` explicite.
    public var liveKitRoom: String { roomId ?? "live-\(id)" }
}
